Add tags to widgets and textboxes

imenattatra opened the pull request:
## What type of PR is this? 
- [x] Feature

## Description

This change adds the possibility to filter the widgets to show according to their tags ,still with the help of an `allowed_widgets` query results that has for each parameter value , the list of allowed tags.

![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExMHJmNmxjN3lhd2lhZDZucHJ5cGswbDAxeno2bXV6MXhrNmcwaHFxbSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/9tT0s3SBnDm4o/giphy.gif)

This was introduced by doing changes both in backend and in frontend. 

**In backend** : 

- We updated the database by introducing the `metrWidget` table which is related to the widget with a 1 to 1 relationship , this allows us to avoid conflicts in the future, the added `metrWidget` has the list of tags of the widget.

- We added an endpoint that allows to save the tags, while saving we check if the widget does not have yet a `metrWidget` we create on the fly, this way old existing widget won't have an issue.

- Even if the used url sends the widget id (`widget_id`), we are having it as` api/metrwidgets ...` to avoid conflicts

**In frontend** : 

 - we  keep the usage of `allowed_widgets_{dashboard_id}`
 - we update the widget component to have the tags (widgets with visualization or textboxes )
 - we update the widget service by adding the call that updates the tags related to a widget
 - we show the tags control in the edit mode only (while editing a dashboard) 

## How is this tested?

- [x] Unit tests (pytest, jest)
- [x] Manually


## Related Tickets & Documents

This PR closes https://github.com/metr-systems/backlog/issues/3436

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

This has been deployed to staging and accessible here https://dashboard.staging.metr.systems/staging/

![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExaGpkeTN6Z2lzdDJhcTFoaDQ2a2txMWszcGo0ZjJjM211c216ejJrdSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/l0IsIWQVQe8UzJT1u/giphy.gif)

imenattatra committed:
add metrWidget in a 1-to-1 relationship with Widget and having the list of tags

imenattatra committed:
create a metrwidget whenever we create a new widget (either with visualization or textbox)

imenattatra committed:
add metrWidgetTagsResource and add its endpoints to the api

imenattatra committed:
add WidgetTagsControl and use it to controld every widget tags

imenattatra committed:
fix missing closing curly bracket

imenattatra committed:
move ´metrWidgetTagsResource´ to a separate file and add tests for it

imenattatra committed:
add test to make sure a metrwidget object is created with every new widget

imenattatra committed:
update ´getAllowedWidgetsForCurrentParam´ to use widget tags

imenattatra committed:
fix getting the current values of the params                                                                                                       Since in the beginning the value is in getParametersDefs() and then it is always in updatedParameters but not always in getParametersDefs()

imenattatra committed:
enhance code in getAllowedWidgetsForCurrentParam

imenattatra committed:
edit widget tags only on edit mode

imenattatra committed:
fix used parameter when checking for allowed widget

this fix makes the filtering work even if we refresh the page

imenattatra committed:
show widget tags only in editing mode

imenattatra committed:
make sure widget tags are updated

imenattatra committed:
remove not needed code

- there is no need to fetch the tags from backend since we are already sending the tags for each widget with the full dashboard source

imenattatra committed:
move axios call to the widget service and remove async usage

imenattatra committed:
create metrwidget if does not already exist when setting its tags

imenattatra committed:
Revert "create a metrwidget whenever we create a new widget (either with visualization or textbox)"

This reverts commit 8153e3700068b2fc5253fedf38bc44e64e94ead3.

imenattatra committed:
remove not used import

imenattatra committed:
rename url using metrwidget instead of widget to avoid conflict

imenattatra committed:
remove not needed imports, variable and tooltip

imenattatra committed:
update tests calling the previsouly renamed metrwidgets url

imenattatra committed:
fix tags to return in case there is no metr_widget attached to the widget

imenattatra committed:
fix initialize archivedTooltip to null rather than removing it

imenattatra committed:
remove not used code

imenattatra committed:
add some styling to the widget controld component

charinass commented on redash/handlers/metr_widgets.py:
```diff
@@ -0,0 +1,37 @@
+)
+from redash.serializers import serialize_widget
+
+class metrWidgetTagsResource(BaseResource):
```
Awesome 💖

charinass commented on tests/handlers/test_metr_widgets.py:
```diff
@@ -0,0 +1,46 @@
+from tests import BaseTestCase
+
+
+class TestMetrWidgetTagsResource(BaseTestCase):
```
💖

charinass reviewed (approved):
Wonderful effort Imen!💖

thiagogds commented on redash/serializers/__init__.py:
```diff
@@ -166,6 +166,7 @@ def serialize_widget(object):
         "text": object.text,
         "updated_at": object.updated_at,
         "created_at": object.created_at,
+        "tags": object.metr_widget[0].tags if object.metr_widget else [],
```
When using a 1on1 to relashioship I would assume that it's possible to use `object.metr_widget.tags` without the `[0]`.

In the [documentation](https://docs.sqlalchemy.org/en/13/orm/basic_relationships.html#one-to-one) it looks like the `uselist=False` is a parameter of the `backref` function and you're passing it to the `relationship` method. Maybe that's the problem

imenattatra replied:
thanks a lot for your feedback thiago , yes i fixed it [here in this commit](https://github.com/metr-systems/redash/pull/4/commits/dde3bb8293de5e68f3bb2b8a313f83c158e51f9a)

thiagogds commented on redash/models/__init__.py:
```diff
@@ -1210,7 +1210,15 @@ def copy(self, dashboard_id):
+class metrWidget(TimestampMixin, BelongsToOrgMixin, db.Model):
+    id = primary_key("metrWidget")
+    widget_id = Column(key_type("Widget"), db.ForeignKey("widgets.id"))
+    widget = db.relationship(Widget,  backref=backref("metr_widget", cascade="delete"),uselist=False)
```
To reference my other comment about the relationship in the serializer. Looks like you should do this:

```suggestion
    widget = db.relationship(Widget,  backref=backref("metr_widget", cascade="delete", ,uselist=False))
```

imenattatra replied:
@thiagogds Thanks a lot for your feedback 

Actually, I tried also that and it was not working but I believe that I was using it the wrong way.

...because i did not want to touch the `Widget` class and apparently I have to, because according to [ the documentation](https://docs.sqlalchemy.org/en/13/orm/basic_relationships.html#one-to-one) , the way we link the models to each other should happen as follow:
- add a `relationship` in one , defining this `uselist` thing (should be done `Widget` table here to avoid any migration for it because relationship alone does not provoke any change in model migration)
- add a `relationship` and a `foreign key` in the other one (should be done in `metrWidget`)

It seems like sqlalchemy does not enforce the definition of the `relationship` in the first model (Widget) but not defining it does not seem correct and i am afraid it would affect how we fetch data in the futur so i will add it.

I imagine that by fixing this... all our issues related to how these two models relate will be fixed...change in progress

imenattatra replied:
Update:
Sorry, Apparently `bacref` in `db.relationship` can be used from one side but `back_populates` needs to be defined from both side , so i am reconsidering what i said earlier and collecting more information

imenattatra replied:
thanks, please find update in [this same commit mentioned previously](https://github.com/metr-systems/redash/pull/4/commits/dde3bb8293de5e68f3bb2b8a313f83c158e51f9a)

thiagogds commented on redash/handlers/api.py:
```diff
@@ -197,6 +198,7 @@ def json_representation(data, code, headers=None):
 
 api.add_org_resource(QueryTagsResource, "/api/queries/tags", endpoint="query_tags")
 api.add_org_resource(DashboardTagsResource, "/api/dashboards/tags", endpoint="dashboard_tags")
+api.add_org_resource(metrWidgetTagsResource, "/api/metrwidgets/<widget_id>/tags", endpoint="widgets_tags")
```
For the sake of pep8, the class name must start with a capital letter. Maybe too much javascript/python context change? hahaha

imenattatra replied:
thanks yeah it was js effect :D , please find update[ in this commit](https://github.com/metr-systems/redash/pull/4/commits/8c7f5078b424aa0f46ff997bcfd71670907c13dd)

thiagogds commented on redash/handlers/metr_widgets.py:
```diff
@@ -0,0 +1,34 @@
+        tags = widget_properties["tags"]
+
+        metrWidget=models.metrWidget.query.filter(models.metrWidget.widget==widget).first()
+        if metrWidget is None:
+            metrWidget=models.metrWidget(widget=widget, tags=tags)
+            models.db.session.add(metrWidget)
+        else:
+            metrWidget.tags = tags
+
```
I'm not sure if this would work because I don't know SQLAlchemy a lot, but I think it would be easier to read without the else:

```suggestion
        if metrWidget is None:
            metrWidget=models.metrWidget(widget=widget)
            models.db.session.add(metrWidget)
        
        metrWidget.tags = tags
```

imenattatra replied:
Thanks, I believe that it is much more readable after i updated it[ in this commit](https://github.com/metr-systems/redash/pull/4/commits/fa3d1d79a3236ea521c555fd268cc62907dd697d), i don't really feel very comfortable removing the if-else because i don't want it to affect the tags twice to the instance but i honestly don't know if it is better than keeping the condition, what do you think about leaving it like that?

imenattatra replied:
thanks, this update is included[ in this commit ](https://github.com/metr-systems/redash/pull/4/commits/7f337577ff4cca5ee75b2645d8884daf85ab7abd)

thiagogds commented on redash/handlers/metr_widgets.py:
```diff
@@ -0,0 +1,34 @@
+        widget_properties = request.get_json(force=True)
+        tags = widget_properties["tags"]
+
+        metrWidget=models.metrWidget.query.filter(models.metrWidget.widget==widget).first()
+        if metrWidget is None:
```
Instead of doing the query manually, can't you use the back relation?

```suggestion
        if not widget.metrWidget: # This is assuming that you can use a 1to1 relation. Otherwise you have to keep the is None
```

imenattatra replied:
thanks, the update is also included[ in this commit ](https://github.com/metr-systems/redash/pull/4/commits/7f337577ff4cca5ee75b2645d8884daf85ab7abd)

thiagogds commented on tests/handlers/test_metr_widgets.py:
```diff
@@ -0,0 +1,33 @@
+from redash import models
+from tests import BaseTestCase
+
+class TestMetrWidgetTagsResource(BaseTestCase):
```
Can you add tests for the permissions?

imenattatra replied:
thanks, please find tests added [here in this commit ](https://github.com/metr-systems/redash/pull/4/commits/9dbea40990cc4f323ce4be3162505e40fe7acc3b)

thiagogds reviewed (approved):
Super cool! Amazing job extending Redash.

I like how you add the new tags to the edition of the dashboard, it feels very easy to use.

I've added a few comments to the python code, but because I'm not used to SQLAlchemy they might be wrong, so I'm already approving the PR

imenattatra committed:
use capital letters for the sake of pep8

imenattatra committed:
[FIX] put uselist param inside tha backward reference

Co-authored-by: Thiago Garcia da Silva <thiago.garciadasilva@metr.systems>

imenattatra committed:
update endpoint that edis metrwidget tags

- use filter_by instead of filter since it is more readable and we have no need of extra flexibility here
- get rid of the else condition

thiagogds commented on redash/handlers/metr_widgets.py:
```diff
@@ -0,0 +1,34 @@
+        widget_properties = request.get_json(force=True)
+        tags = widget_properties["tags"]
+
+        metrWidget = models.metrWidget.query.filter_by(widget=widget).first()
+        if metrWidget:
+            metrWidget.tags = tags
+        else:
+            metrWidget = models.metrWidget(widget=widget, tags=tags)
+            models.db.session.add(metrWidget)
+
+        models.db.session.commit()
```
You must always add the `metrWidget` to the DB session because you also changed the tags in the if. 
Also, in the serializer at `redash/serializers/__init__.py` you use `object.metr_widget`, so I think it should be possible here too.

I also changed it to be snake_case instead of camelCase ;) 

``` suggestion
        metr_widget = widget.metr_widget
        if metr_widget:
            metr_widget.tags = tags
        else:
            metr_widget = models.metrWidget(widget=widget, tags=tags)
            
        models.db.session.add(metr_widget)
        models.db.session.commit()
```

imenattatra replied:
I was checking yesterday if i do really need to add it again to the session and the documentation was saying that we only add it if new or not already in session, so i used `inspect` 
```
from sqlalchemy import inspect

state = inspect(metr_widget)
if state.persistent:
    print("The object is persistent and in the session.")
```
and it seemed to be already in the session but i was confused because i saw in other code in redash they also add it again to the session after an update.

So i believe that yes i will add it again just in case

imenattatra replied:
please find the update [here in this commit ](https://github.com/metr-systems/redash/pull/4/commits/7f337577ff4cca5ee75b2645d8884daf85ab7abd)

thiagogds replied:
It might as well be this and people just don't feel secure enough and add it again hehe

thiagogds reviewed (changes requested):
I made a few suggestion to get the code easier to read, but the main change is related to the db session

imenattatra committed:
use widget.metr_widget directly

imenattatra committed:
update adding tags to widget code
- rename metrWidget to metr_widget
- always add metr_widget to session before commiting

thiagogds reviewed (approved):
Awesome work! Thanks for looking into my suggestions even though I wasn't so sure they were possible :)

I'm already approving, but would be cool to add the tests I mentioned before

imenattatra committed:
add permission tests for updating widget related tags

imenattatra commented:
> Awesome work! Thanks for looking into my suggestions even though I wasn't so sure they were possible :)
> 
> I'm already approving, but would be cool to add the tests I mentioned before

Thanks, i added the tests[ here in the latest commit ](https://github.com/metr-systems/redash/pull/4/commits/9dbea40990cc4f323ce4be3162505e40fe7acc3b)

imenattatra commented:
Just as a note here: 

I was digging deeper into making the permissions tests better , my plan was to have something similar to the following (or more detailed) in order to have track exactly of the permissions checked

```python
  def test_post_allowed_if_user_has_permission_to_edit_dashboard(self):
      # in other words if admin user
      widget = self.factory.create_widget()
      user = self.factory.create_admin(org=self.factory.org,group_ids=[self.factory.admin_group.id, self.factory.default_group.id]) # admin user

      self.assertTrue(user.has_permission("edit_dashboard"))
      self.assertTrue(can_modify(widget.dashboard, user))


      response = self.make_request("post",f"/api/metrwidgets/{widget.id}/tags",data= {"tags":["tag1", "tag2"]}, user=user)
      self.assertEqual(response.status_code, 200)
```
but since the authentication happens inside the `self.make_request` method, `can_modify` fails if i try to check on it earlier
